src/goto-checker/goto_symex_property_decider.cpp

```diff
@@ -131,27 +132,40 @@ void goto_symex_property_decidert::update_properties_status_from_goals(
   case decision_proceduret::resultt::D_SATISFIABLE:
     for(auto &goal_pair : goal_map)
     {
-      auto &status = properties.at(goal_pair.first).status;
+      property_infot &property = properties.at(goal_pair.first);
+      auto &status = property.status;
       if(
         solver->decision_procedure()
           .get(goal_pair.second.condition)
           .is_true() &&
         status != property_statust::FAIL)
       {
-        status |= property_statust::FAIL;
+        if(property.pc->is_expected_valid())
```

Comment:
I spent a long time disentangling and vertically slicing the implementations of the different semantics of coverage and assertions into different goto_verifiert/goto_checkert implementations to simplify the logic for the various modes in which CBMC can be run. This seems like a step backwards, mixing presentation layer issues into low level logic. At this level there should only be one semantics of assertions. If you want to present them differently then this needs to be handled at a higher level, e.g. in a new goto_verifier implementation. I haven't fully thought this through, but this is my first intuition.

Reply:
I agree with Peter's view.